### ISSUE TITLE
libvirt_volume.main_disk: Can't retrieve volume xxx

### ISSUE THREAD

**Issue by MalloZup** [dmacvicar/terraform-provider-libvirt#88]
## How-to-reproduce:

clone sumaform
Take the main.tf-libvirt-testing example.
```console
Terraform get
Terraform apply
```
run it 3/4 times.  ( 2 times will fail with error : 
```
libvirt_volume.main_disk: Can't retrieve volume  sles12sp1
```
This issue block the provissioning and for production or dayly work is annoyng, because we have to reapply terraform apply.

I cannot sadly determine the cause of this issue, the disk created is there but terraform fails to retrieve it. 
It's a race condition. I'm quite sure that the sumaform is not the root cause of the problem.

For additional logs, info, ping me

**Comment by MalloZup** [comment]
Doing 2 times terraform apply workarounds the problem

**Comment by MalloZup** [comment]
@inercia  @moio 
here we have the log with debug enabled:

https://gist.github.com/MalloZup/0413641e9f350d22f225f0982f577fef

Looking at this debug, to me it looks like the issue is that terraform want to use the volume, but the volume is not there, and the plugin says "it's created".


cloning ```https://github.com/moio/sumaform```
Here what i used  as tf.file

```console 
provider "libvirt" {
  uri = "qemu+tcp://myREMOTE_SERVER_Desperados.server.com/system"
}

module "base" {
  source = "./modules/libvirt/base"

  cc_username = "UC7"
  cc_password = "Desperados"

  pool = "ssd"
  network_name = ""
  bridge = "br0"
  name_prefix = "dg"
  use_avahi = false
  domain = "mgr.suse.de"
  images = ["sles12sp2"]
}

module "vanilla" {
  source = "./modules/libvirt/host"
  base_configuration = "${module.base.configuration}"
  name = "vanilla"
  image = "sles12sp2"
}
```

So here we create the disks, globally, https://github.com/moio/sumaform/blob/master/modules/laibvirt/base/main.tf#L5


then for the host we use them here:

https://github.com/moio/sumaform/blob/master/modules/libvirt/host/main.tf

**Comment by MalloZup** [comment]
this is a invalid issue, due to uncorrect use of the plugin. thx for support, closing

**Comment by moio** [comment]
> this is a invalid issue, due to uncorrect use of the plugin. thx for support, closing

Why? Did you find a workaround? I now have the same problem.

**Comment by MalloZup** [comment]
@moio  was at the time we did not lock resource esplictely via hcl
I think i have seen it also time ago, we might repoen it. but  i fear also is on libvirt domain design also

**Comment by ktsamis** [comment]
This needs fixing because the jenkins jobs randomly fail. I propose a modification of the [script here](https://gitlab.suse.de/galaxy/sumaform-test-runner/blob/master/jenkins-runner.sh#L23) to check if it failed to try again to rerun the command and then if that fails again, then fail the test setup. As soon as I have some time I will open a PR for that but it will be a workaround.

**Comment by MalloZup** [comment]
@ktsamis  this GitHub project is the the "core" terraform-plugin. Sumaform or whatever is only a secondary way how to use the plugin. The main issue here is about the libvirt volume, which impact sumaform and others
Feel free to discuss on that PR on sumaform-test-runner repo for the workaround. :+1: 

ciao

**Comment by ktsamis** [comment]
Here I mostly wanted to say that the issue affects us still, and at a quite large degree.

**Comment by bpetit** [comment]
I'm not completly sure it is the same bug, but I encountered something very similar on both destroy and apply:


> terraform destroy
> libvirt_volume.test-disk4: Refreshing state... (ID: /var/lib/libvirt/images/test-disk4)
> libvirt_volume.test-disk3: Refreshing state... (ID: /var/lib/libvirt/images/test-disk3)
> libvirt_network.test-net: Refreshing state... (ID: e673c11c-821e-4e71-8d8b-f7ee75212c3f)
> libvirt_volume.test-disk5: Refreshing state... (ID: /var/lib/libvirt/images/test-disk5)
> libvirt_volume.test-disk2: Refreshing state... (ID: /var/lib/libvirt/images/test-disk2)
> libvirt_volume.test-disk: Refreshing state... (ID: /var/lib/libvirt/images/test-disk)
> libvirt_volume.test-disk6: Refreshing state... (ID: /var/lib/libvirt/images/test-disk6)
> libvirt_domain.vm04: Refreshing state... (ID: 70e4384c-2238-4a0c-9152-e97586b92cc4)
> libvirt_domain.vm03: Refreshing state... (ID: d538c95d-f9e4-4600-9025-06c1efc36048)
> libvirt_domain.vm01: Refreshing state... (ID: bc556ae2-ecb3-4d4a-9610-e1b243eeae4e)
> 
> Error: Error refreshing state: 3 error(s) occurred:
> 
> * libvirt_volume.test-disk5: 1 error(s) occurred:
> 
> * libvirt_volume.test-disk5: libvirt_volume.test-disk5: can't retrieve volume /var/lib/libvirt/images/test-disk5
> * libvirt_volume.test-disk6: 1 error(s) occurred:
> 
> * libvirt_volume.test-disk6: libvirt_volume.test-disk6: can't retrieve volume /var/lib/libvirt/images/test-disk6
> * libvirt_volume.test-disk2: 1 error(s) occurred:
> 
> * libvirt_volume.test-disk2: libvirt_volume.test-disk2: can't retrieve volume /var/lib/libvirt/images/test-disk2
> 

> 
> terraform apply 
> libvirt_volume.test-disk6: Refreshing state... (ID: /var/lib/libvirt/images/test-disk6)
> libvirt_volume.test-disk5: Refreshing state... (ID: /var/lib/libvirt/images/test-disk5)
> libvirt_volume.test-disk2: Refreshing state... (ID: /var/lib/libvirt/images/test-disk2)
> libvirt_volume.test-disk4: Refreshing state... (ID: /var/lib/libvirt/images/test-disk4)
> libvirt_volume.test-disk3: Refreshing state... (ID: /var/lib/libvirt/images/test-disk3)
> libvirt_volume.test-disk: Refreshing state... (ID: /var/lib/libvirt/images/test-disk)
> libvirt_network.test-net: Refreshing state... (ID: e673c11c-821e-4e71-8d8b-f7ee75212c3f)
> libvirt_domain.vm05: Refreshing state... (ID: c8415925-0623-494e-b24f-afd31f238262)
> libvirt_domain.vm01: Refreshing state... (ID: bc556ae2-ecb3-4d4a-9610-e1b243eeae4e)
> libvirt_domain.vm06: Refreshing state... (ID: 213f4eee-7eb1-419d-9c8e-c2c7cfa41f8b)
> 
> Error: Error refreshing state: 3 error(s) occurred:
> 
> * libvirt_volume.test-disk4: 1 error(s) occurred:
> 
> * libvirt_volume.test-disk4: libvirt_volume.test-disk4: can't retrieve volume /var/lib/libvirt/images/test-disk4
> * libvirt_volume.test-disk2: 1 error(s) occurred:
> 
> * libvirt_volume.test-disk2: libvirt_volume.test-disk2: can't retrieve volume /var/lib/libvirt/images/test-disk2
> * libvirt_volume.test-disk3: 1 error(s) occurred:
> 
> * libvirt_volume.test-disk3: libvirt_volume.test-disk3: can't retrieve volume /var/lib/libvirt/images/test-disk3`
> 

The volumes do exist:

> 
> virsh vol-list default
>  Name                 Path                                    
> ------------------------------------------------------------------------------
>  test                 /var/lib/libvirt/images/test            
>  test-disk            /var/lib/libvirt/images/test-disk       
>  test-disk2           /var/lib/libvirt/images/test-disk2      
>  test-disk3           /var/lib/libvirt/images/test-disk3      
>  test-disk4           /var/lib/libvirt/images/test-disk4      
>  test-disk5           /var/lib/libvirt/images/test-disk5      
>  test-disk6           /var/lib/libvirt/images/test-disk6      
>  test2                /var/lib/libvirt/images/test2           
>  test3                /var/lib/libvirt/images/test3       

> 
> ls /var/lib/libvirt/images/
> test  test2  test3  test-disk  test-disk2  test-disk3  test-disk4  test-disk5  test-disk6

Volumes source files are present too.

Let me know if I can paste more useful data.

**Comment by Bischoff** [comment]
It is worse with terraform 11.7 and the provider 0.4.2: it now happens on *all* volumes and *all* the times.

An "easy but dirty" workaround is to remove all volumes manually with a simple `rm` command after the `terraform destroy`.

**Comment by moio** [comment]
Does `terraform apply -parallelism=1` work as it did with older Terraforms?

**Comment by Bischoff** [comment]
With `terraform destroy -parallelism=1` the destroy at least completes, but now we have slightly different messages:
```
libvirt_volume.main_disk: Can't retrieve volume /home/ebischoff/KVM_VMs/ebi3-proxy-main-disk
 XML desc: virError(Code=1, Domain=0, Message='An error occurred, but the cause is unknown')
```

**Comment by moio** [comment]
Are the disks really gone (ie. we could disregard "an error occurred, but the cause is unknown")?

**Comment by Bischoff** [comment]
No, they are still there, but the domains are really gone.

So a workaround could be: `terraform destroy` with `-parallelism=1`, ignore errors and the destroy volumes manually.

**Comment by Bischoff** [comment]
The more I think to it, the more I'm convinced we have two distinct problems:
* the race condition, worked around with `-parallelism=1`
* the libvirt problem

If that's so, sorry to have hijacked this issue for the second problem.

**Comment by MalloZup** [comment]
To all:

Resuming this issue was about retrieve volumes by id.

As user we get the error:
`libvirt_volume.main_disk: Can't retrieve volume  sles12sp1`


As @dmacvicar  stated in his PR comment, at the end  
` and probably (no evidence) related ones like #324 and #88`

https://github.com/dmacvicar/terraform-provider-libvirt/pull/353#issue-207733280

Since this issue was a race condition we cannot test with evidence.

I'm an optimistic person, so i will close this issue. 

To all sumaform users @moio @ktsamis @Bischoff : until we don't update and merge this https://github.com/SUSE/spacewalk/issues/5475 

is clear that you will continue to have or experiment this issues.

If you have again the error **after** update,  and i mean the error cannot retrieve volume that i post above, feel free to reopen this issue.

Optimal would be that you run `TF_LOG=DEBUG terraform apply` so we have full debug.

I will thnk if we can maybe enable TF_LOG=DEBUG in production and saving maybe only the debug info in a special file

**Comment by MalloZup** [comment]
https://github.com/SUSE/spacewalk/issues/5544

**Comment by moio** [comment]
This is actually confirmed to be a core libvirt issue fixed from version 4.3.0 on. Specifically by this patch, that can be applied back to 4.0.0 without issues:

https://bugzilla.redhat.com/show_bug.cgi?id=1576464#c3

**Comment by moio** [comment]
Reported against Ubuntu as: https://bugs.launchpad.net/ubuntu/+source/libvirt/+bug/1789659
Reported against SUSE distros as: https://bugzilla.suse.com/show_bug.cgi?id=1106420

openSUSE Tumbleweed is not affected
